File: defence_claw/defenseclaw/cli/defenseclaw/scanner/skill.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Skill scanner — native SDK integration.

Uses the cisco-ai-skill-scanner Python SDK directly instead of shelling out
to the skill-scanner CLI.  Maps SDK ScanResult/Finding → DefenseClaw models.
"""

from __future__ import annotations

import logging
import os
import sys
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING

from defenseclaw.config import (
    CiscoAIDefenseConfig,
    InspectLLMConfig,
    LLMConfig,
    SkillScannerConfig,
)
from defenseclaw.models import Finding, ScanResult
from defenseclaw.scanner._llm_env import (
    inject_llm_env,
    litellm_model,
    llm_analyzer_ready,
)

if TYPE_CHECKING:
    pass

_log = logging.getLogger(__name__)


def _inspect_to_llm(il: InspectLLMConfig) -> LLMConfig:
    """Back-compat shim — mirrors the one in ``mcp.py``. Kept local so
    each scanner can be deleted independently when we fully retire the
    legacy ``InspectLLMConfig`` shape."""
    return LLMConfig(
        model=il.model,
        provider=il.provider,
        api_key=il.api_key,
        api_key_env=il.api_key_env,
        base_url=il.base_url,
        timeout=il.timeout,
        max_retries=il.max_retries,
    )


class SkillScannerWrapper:
    """Wraps the cisco-ai-skill-scanner SDK.

    Mirrors :class:`MCPScannerWrapper` — accepts either a legacy
    ``InspectLLMConfig`` or a unified ``LLMConfig`` via ``llm=``.
    Internally everything is driven through :class:`LLMConfig` and the
    shared :mod:`defenseclaw.scanner._llm_env` helpers.
    """

    def __init__(
        self,
        config: SkillScannerConfig,
        inspect_llm: InspectLLMConfig | None = None,
        cisco_ai_defense: CiscoAIDefenseConfig | None = None,
        *,
        llm: LLMConfig | None = None,
    ) -> None:
        self.config = config
        self.inspect_llm = inspect_llm or InspectLLMConfig()
        self.cisco_ai_defense = cisco_ai_defense or CiscoAIDefenseConfig()
        self._llm: LLMConfig = llm if llm is not None else _inspect_to_llm(self.inspect_llm)

    def name(self) -> str:
        return "skill-scanner"

    def scan(self, target: str) -> ScanResult:
        import time

        try:
            from skill_scanner import SkillScanner
            from skill_scanner.core.analyzer_factory import build_analyzers
            from skill_scanner.core.scan_policy import ScanPolicy
        except ImportError:
            print(
                "error: cisco-ai-skill-scanner not installed.\n"
                "  Repair the managed DefenseClaw installation.\n"
                "  Source checkouts: uv sync",
                file=sys.stderr,
            )
            raise SystemExit(1)

        cfg = self.config
        llm = self._llm
        self._inject_env()

        policy = ScanPolicy.default()
        if cfg.policy:
            try:
                policy = ScanPolicy.from_file(cfg.policy)
            except Exception:
                presets = {"strict", "balanced", "permissive"}
                if cfg.policy in presets:
                    policy = ScanPolicy.from_preset(cfg.policy)

        build_kwargs: dict = {"policy": policy}
        if cfg.use_behavioral:
            build_kwargs["use_behavioral"] = True
        if cfg.use_llm:
            # The upstream skill-scanner SDK auto-detects the provider
            # from a LiteLLM-shaped ``provider/model`` string via its
            # ``ProviderConfig`` (Bedrock, Gemini, Vertex, Azure,
            # Ollama, OpenRouter, …). We deliberately do NOT pass
            # ``llm_provider`` here because:
            #   1. The factory ignores it whenever ``llm_model`` is
            #      set (skill_scanner/core/analyzer_factory.py).
            #   2. Our internal short names ("bedrock", "vertex_ai")
            #      don't match the upstream ``LLMProvider`` enum
            #      ("aws-bedrock", "gcp-vertex"), so passing them
            #      would only matter on the model-less path and
            #      would error out there.
            # Letting the model string carry the provider keeps every
            # LiteLLM-supported provider working end-to-end.
            #
            # Guard ``use_llm`` on a resolved model: without it, the
            # upstream factory falls back to a hard-coded Anthropic
            # default (``claude-3-5-sonnet-20241022``) and then crashes
            # on operators whose unified key isn't an Anthropic key.
            # Skipping the LLM analyzer with a clear log line is
            # strictly better than emitting an upstream warning that
            # operators can't action.
            model = litellm_model(llm)
            env_model = os.environ.get("SKILL_SCANNER_LLM_MODEL", "")
            effective_model = model or env_model
            api_key = llm.resolved_api_key() or os.environ.get(
                "SKILL_SCANNER_LLM_API_KEY", ""
            )
            if effective_model and llm_analyzer_ready(
                llm,
                model=effective_model,
                api_key=api_key,
            ):
                build_kwargs["use_llm"] = True
                if model:
                    build_kwargs["llm_model"] = model
                elif env_model:
                    build_kwargs["llm_model"] = env_model
                if api_key:
                    build_kwargs["llm_api_key"] = api_key
                if llm.base_url:
                    build_kwargs["llm_base_url"] = llm.base_url
                if cfg.llm_consensus_runs > 0:
                    build_kwargs["llm_consensus_runs"] = cfg.llm_consensus_runs
            elif effective_model:
                key_name = llm.api_key_env or "DEFENSECLAW_LLM_KEY"
                print(
                    "warning: LLM analyzer skipped: "
                    f"{key_name} is not configured; continuing with local analyzers",
                    file=sys.stderr,
                )
            else:
                _log.info(
                    "skill-scanner: use_llm requested but no model resolved "
                    "from llm.model / SKILL_SCANNER_LLM_MODEL — skipping LLM "
                    "analyzer to avoid upstream's Anthropic fallback default. "
                    "Set llm.model (e.g. 'bedrock/anthropic.claude-3-5-haiku') "
                    "or SKILL_SCANNER_LLM_MODEL to enable.",
                )
        if cfg.use_trigger:
            build_kwargs["use_trigger"] = True
        if cfg.use_virustotal:
            build_kwargs["use_virustotal"] = True
        if cfg.use_aidefense:
            build_kwargs["use_aidefense"] = True

        analyzers = build_analyzers(**build_kwargs)
        scanner = SkillScanner(analyzers=analyzers, policy=policy)

        start = time.monotonic()
        sdk_result = scanner.scan_skill(str(target), lenient=cfg.lenient)
        elapsed = time.monotonic() - start

        return self._convert(sdk_result, target, elapsed)

    def _inject_env(self) -> None:
        """Inject API keys and the skill-scanner-specific env vars.

        Two layers:

        1. Provider-specific env vars for LiteLLM (via the shared
           helper). This is how the analyzer eventually reaches the
           model regardless of provider.
        2. skill-scanner's bespoke env vars (``SKILL_SCANNER_LLM_*``,
           ``VIRUSTOTAL_API_KEY``, ``AI_DEFENSE_API_KEY``) that the SDK
           reads directly. Kept here until skill-scanner switches to the
           provider-native env vars.
        """
        cfg = self.config
        llm = self._llm
        aid = self.cisco_ai_defense
        inject_llm_env(llm)

        mappings = [
            ("SKILL_SCANNER_LLM_API_KEY", llm.resolved_api_key()),
            ("SKILL_SCANNER_LLM_MODEL", litellm_model(llm)),
            ("VIRUSTOTAL_API_KEY", cfg.resolved_virustotal_api_key()),
            ("AI_DEFENSE_API_KEY", aid.resolved_api_key()),
        ]
        for env_var, value in mappings:
            if value and env_var not in os.environ:
                os.environ[env_var] = value

    def _convert(self, sdk_result: object, target: str, elapsed: float) -> ScanResult:
        """Convert SDK ScanResult → DefenseClaw ScanResult."""
        scanner_name = self.name()
        findings: list[Finding] = []
        for sf in getattr(sdk_result, "findings", []):
            location = getattr(sf, "file_path", "") or ""
            line = getattr(sf, "line_number", None)
            if line and location:
                location = f"{location}:{line}"

            tags: list[str] = []
            category = getattr(sf, "category", None)
            if category:
                cat_name = category.name if hasattr(category, "name") else str(category)
                tags.append(cat_name)
            analyzer = str(getattr(sf, "analyzer", "") or "")
            if analyzer:
                tags.append(f"analyzer:{analyzer}")

            severity = getattr(sf, "severity", None)
            sev_str = severity.name if hasattr(severity, "name") else str(severity)

            findings.append(Finding(
                id=getattr(sf, "id", "") or getattr(sf, "rule_id", ""),
                severity=sev_str.upper(),
                title=getattr(sf, "title", ""),
                description=getattr(sf, "description", ""),
                location=location,
                remediation=getattr(sf, "remediation", "") or "",
                # Canonical finding identity names the producer, not the
                # upstream SDK's internal analyzer, which remains in tags.
                scanner=scanner_name,
                tags=tags,
            ))

        return ScanResult(
            scanner=scanner_name,
            target=target,
            timestamp=datetime.now(timezone.utc),
            findings=findings,
            duration=timedelta(seconds=elapsed),
        )

File: defence_claw/defenseclaw/internal/scanner/skill.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package scanner

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/defenseclaw/defenseclaw/internal/config"
	"github.com/defenseclaw/defenseclaw/internal/processutil"
)

var ansiRe = regexp.MustCompile(`\x1b\[[0-9;]*[a-zA-Z]`)

// liteLLMModel returns the model string shaped for LiteLLM /
// provider-native routers. LiteLLM accepts “"provider/model-id"“
// directly — the same shape DefenseClaw uses in config. A bare
// “llm.Model“ with a separate “llm.Provider“ gets stitched into
// “"<provider>/<model>"“. Empty models are passed through unchanged
// (the caller should handle that case).
func liteLLMModel(llm config.LLMConfig) string {
	model := llm.Model
	if model != "" && llm.Provider != "" && !strings.Contains(model, "/") {
		return llm.Provider + "/" + model
	}
	return model
}

// extractJSON finds the first top-level JSON object in data.
// Scanner CLIs sometimes print progress text to stdout before the JSON;
// this isolates the `{...}` payload so json.Unmarshal succeeds.
// extractJSON locates the first balanced JSON object in data by tracking
// brace depth while skipping string literals.
func extractJSON(data []byte) []byte {
	start := bytes.IndexByte(data, '{')
	if start < 0 {
		return data
	}
	depth := 0
	inString := false
	escaped := false
	for i := start; i < len(data); i++ {
		b := data[i]
		if escaped {
			escaped = false
			continue
		}
		if b == '\\' && inString {
			escaped = true
			continue
		}
		if b == '"' {
			inString = !inString
			continue
		}
		if inString {
			continue
		}
		switch b {
		case '{':
			depth++
		case '}':
			depth--
			if depth == 0 {
				return data[start : i+1]
			}
		}
	}
	return data
}

// SkillScanner shells out to the Python “cisco-ai-skill-scanner“ CLI.
//
// The LLM-facing surface is driven by the unified “config.LLMConfig“
// (top-level “llm:“ merged with “scanners.skill.llm:“ overrides,
// resolved once by “Config.ResolveLLM“). “InspectLLM“ is kept as a
// deprecated back-compat field: old callers that constructed the
// scanner with an “InspectLLMConfig“ still work because
// “NewSkillScanner“ translates it into “LLM“ on the way in. New
// call sites should use “NewSkillScannerFromLLM“ and pass the
// resolved unified config directly.
type SkillScanner struct {
	Config         config.SkillScannerConfig
	LLM            config.LLMConfig
	InspectLLM     config.InspectLLMConfig // Deprecated: populated only for back-compat; do not read.
	CiscoAIDefense config.CiscoAIDefenseConfig
}

// inspectToLLM copies the legacy “InspectLLMConfig“ shape into the
// unified “LLMConfig“ so the scanner can drive everything off a
// single structure internally. Kept local so “config“ doesn't grow
// another conversion helper for a shape we plan to retire.
func inspectToLLM(il config.InspectLLMConfig) config.LLMConfig {
	return config.LLMConfig{
		Model:      il.Model,
		Provider:   il.Provider,
		APIKey:     il.APIKey,
		APIKeyEnv:  il.APIKeyEnv,
		BaseURL:    il.BaseURL,
		Timeout:    il.Timeout,
		MaxRetries: il.MaxRetries,
	}
}

// NewSkillScanner is the back-compat constructor. Accepts the legacy
// “InspectLLMConfig“ shape and translates it into the unified
// “LLMConfig“ internally so downstream code only needs one path.
// New callers should use “NewSkillScannerFromLLM“ and pass
// “Config.ResolveLLM("scanners.skill")“ directly.
func NewSkillScanner(cfg config.SkillScannerConfig, llm config.InspectLLMConfig, aid config.CiscoAIDefenseConfig) *SkillScanner {
	if cfg.Binary == "" {
		cfg.Binary = "skill-scanner"
	}
	return &SkillScanner{
		Config:         cfg,
		LLM:            inspectToLLM(llm),
		InspectLLM:     llm,
		CiscoAIDefense: aid,
	}
}

// NewSkillScannerFromLLM constructs a scanner directly from the
// resolved unified LLM config. Preferred constructor — call sites
// should resolve once via “rootCfg.ResolveLLM("scanners.skill")“ and
// pass the result here so per-scanner overrides on top of the
// top-level “llm:“ block are honored.
func NewSkillScannerFromLLM(cfg config.SkillScannerConfig, llm config.LLMConfig, aid config.CiscoAIDefenseConfig) *SkillScanner {
	if cfg.Binary == "" {
		cfg.Binary = "skill-scanner"
	}
	return &SkillScanner{
		Config:         cfg,
		LLM:            llm,
		CiscoAIDefense: aid,
	}
}

func (s *SkillScanner) Name() string               { return "skill-scanner" }
func (s *SkillScanner) Version() string            { return "1.0.0" }
func (s *SkillScanner) SupportedTargets() []string { return []string{"skill"} }

func (s *SkillScanner) buildArgs(target string) []string {
	args := []string{"scan", "--format", "json"}

	useLLM := s.Config.UseLLM && skillScannerSupportsLLMProvider(s.LLM.ProviderPrefix())
	if useLLM {
		args = append(args, "--use-llm")
	}
	if s.Config.UseBehavioral {
		args = append(args, "--use-behavioral")
	}
	if s.Config.EnableMeta {
		args = append(args, "--enable-meta")
	}
	if s.Config.UseTrigger {
		args = append(args, "--use-trigger")
	}
	if s.Config.UseVirusTotal {
		args = append(args, "--use-virustotal")
	}
	if s.Config.UseAIDefense {
		args = append(args, "--use-aidefense")
	}
	if useLLM && s.LLM.Provider != "" {
		args = append(args, "--llm-provider", s.LLM.Provider)
	}
	if useLLM && s.Config.LLMConsensus > 0 {
		args = append(args, "--llm-consensus-runs", strconv.Itoa(s.Config.LLMConsensus))
	}
	if s.Config.Policy != "" {
		args = append(args, "--policy", s.Config.Policy)
	}
	if s.Config.Lenient {
		args = append(args, "--lenient")
	}

	args = append(args, target)
	return args
}

func skillScannerSupportsLLMProvider(provider string) bool {
	switch strings.ToLower(strings.TrimSpace(provider)) {
	case "", "anthropic", "openai":
		return true
	default:
		return false
	}
}

// scanEnv returns the process environment with skill-scanner-specific
// API keys injected from config. Values already present in the
// environment are not overwritten.
func (s *SkillScanner) scanEnv() []string {
	env := os.Environ()

	inject := []struct {
		envVar string
		value  string
	}{
		// skill-scanner's bespoke env vars. The underlying Python
		// scanner reads these directly today — we keep writing them
		// until skill-scanner migrates to provider-native env vars.
		// ``LiteLLMModel`` stitches bare model + provider into
		// ``provider/model`` when needed so LiteLLM can route it.
		{"SKILL_SCANNER_LLM_API_KEY", s.LLM.ResolvedAPIKey()},
		{"SKILL_SCANNER_LLM_MODEL", liteLLMModel(s.LLM)},
		{"VIRUSTOTAL_API_KEY", s.Config.ResolvedVirusTotalKey()},
		{"AI_DEFENSE_API_KEY", s.CiscoAIDefense.ResolvedAPIKey()},
	}

	existing := make(map[string]bool)
	for _, e := range env {
		for i := 0; i < len(e); i++ {
			if e[i] == '=' {
				existing[e[:i]] = true
				break
			}
		}
	}

	for _, kv := range inject {
		if kv.value != "" && !existing[kv.envVar] {
			env = append(env, kv.envVar+"="+kv.value)
		}
	}

	if !existing["NO_COLOR"] {
		env = append(env, "NO_COLOR=1")
	}
	if !existing["TERM"] {
		env = append(env, "TERM=dumb")
	}

	return env
}

func (s *SkillScanner) Scan(ctx context.Context, target string) (*ScanResult, error) {
	start := time.Now()
	exitCode := 0
	var scanErr error
	var result *ScanResult

	args := s.buildArgs(target)
	cmd := processutil.CommandContext(ctx, s.Config.Binary, args...)
	cmd.Env = s.scanEnv()

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	duration := time.Since(start)
	stderrStr := stderr.String()

	result = &ScanResult{
		Scanner:    s.Name(),
		Target:     target,
		Timestamp:  start,
		Duration:   duration,
		TargetType: InferTargetType(s.Name()),
	}

	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			exitCode = exitErr.ExitCode()
		}
		if errors.Is(err, exec.ErrNotFound) {
			scanErr = fmt.Errorf("scanner: %s not found at %q — repair the managed DefenseClaw installation; source checkouts: uv sync", s.Name(), s.Config.Binary)
			return nil, scanErr
		}
		if stdout.Len() == 0 {
			scanErr = fmt.Errorf("scanner: %s exited %d: %s", s.Name(), exitCode, stderrStr)
			return nil, scanErr
		}
	}

	result.ExitCode = exitCode
	if exitCode != 0 {
		result.ScanError = stderrStr
	}

	if stdout.Len() > 0 {
		findings, parseErr := parseSkillOutput(stdout.Bytes())
		if parseErr != nil {
			scanErr = fmt.Errorf("scanner: failed to parse %s output: %w (stderr=%s)", s.Name(), parseErr, stderrStr)
			return nil, scanErr
		}
		result.Findings = findings
	}

	// hardening (S2.scanners): fail closed on any non-zero
	// scanner exit even when stdout parsed cleanly. See the matching
	// comment in mcp.go and finding "Non-zero skill scanner exits
	// can be treated as successful scans".
	if exitCode != 0 {
		scanErr = fmt.Errorf("scanner %s exited %d (stderr=%s)", s.Name(), exitCode, stderrStr)
		return result, scanErr
	}

	return result, nil
}

type skillOutput struct {
	Findings []skillFinding `json:"findings"`
}

type skillFinding struct {
	ID          string `json:"id"`
	Severity    string `json:"severity"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Location    string `json:"location"`
	Remediation string `json:"remediation"`
	RuleID      string `json:"rule_id"`
	Category    string `json:"category"`
	Line        int    `json:"line"`
}

func parseSkillOutput(data []byte) ([]Finding, error) {
	clean := extractJSON(ansiRe.ReplaceAll(data, nil))
	var out skillOutput
	if err := json.Unmarshal(clean, &out); err != nil {
		return nil, err
	}

	findings := make([]Finding, 0, len(out.Findings))
	for _, f := range out.Findings {
		var ln *int
		if f.Line > 0 {
			v := f.Line
			ln = &v
		}
		findings = append(findings, Finding{
			ID:          f.ID,
			Severity:    Severity(f.Severity),
			Title:       f.Title,
			Description: f.Description,
			Location:    f.Location,
			Remediation: f.Remediation,
			Scanner:     "skill-scanner",
			RuleID:      f.RuleID,
			Category:    f.Category,
			LineNumber:  ln,
		})
	}
	return findings, nil
}
